### PR TITLE
Refactor FXIOS-9778 Add Multipath TCP (MPTCP) support

### PR DESCRIPTION
**Pull request by Aperence**
## :scroll: Tickets
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/21470)
[Jira](https://mozilla-hub.atlassian.net/browse/FXIOS-9778)

## :bulb: Description

MPTCP is a TCP extension allowing to improve network reliabilty by using mutiple network interface for the same TCP connection. Check https://www.mptcp.dev and https://developer.apple.com/documentation/foundation/urlsessionconfiguration/improving_network_reliability_using_multipath_tcp for details. Changes to this repository includes introducing new configurations (defaultMPTCP/ephemeralMPTCP), replacing uses of default/ephemeral by these new configurations, introducing a sharedMPTCP property on URLSession and modifying the makeURL function to enable the handover mode by default

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [X] Wrote unit tests and/or ensured the tests suite is passing
- [X] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [X] If needed, I updated documentation / comments for complex code and public methods
- [X] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by mattreaganmozilla** [comment]
Thanks for this PR @Aperence. Might be good to get some extra eyes on this (cc @nbhasin2).

File: firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift
Comment (nbhasin2) [inline firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift:24]: conf is ephemeral, not default
```suggestion
        // If we aren't building for iOS, ephemeralMPTCP == ephemeral, thus fallback to TCP
```

File: firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift
Comment (nbhasin2) [inline firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift:19]: nit: capitalization
```suggestion
        var conf = self.ephemeral
```

File: firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift
Comment (nbhasin2) [inline firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift:9]: nit: capitalization
```suggestion
        var conf = self.default
```

**Review by nbhasin2 — APPROVED** [review]
Approving with minor nits

**Comment by nbhasin2** [comment]
image

@Aperence I see the following tests failing, would you mind looking into it and let us know if any support is required

File: firefox-ios/Client/Extensions/URLSession+sharedMPTCP.swift
Comment (nbhasin2) [inline firefox-ios/Client/Extensions/URLSession+sharedMPTCP.swift:1]: Thanks @Aperence, somehow the warnings have increased. Let me re-run the build but I would suggest trying locally to check the warnings between main / your branch 

Just posting in this file to have a chain of comments

Comment (Aperence) [reply firefox-ios/Client/Extensions/URLSession+sharedMPTCP.swift:1]: Hello @nbhasin2,

Thank you for the review first.
After having checked the code, it seems like it was simply 2 immutable variables that were declared with a `var` keyword, instead of `let`. This should be fixed normally.

Also, I've seen that I forgot to add the [`Multipath`](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_networking_multipath) entitlement, I've just added it now (if it is fine ?).

Have a nice day

Comment (nbhasin2) [reply firefox-ios/Client/Extensions/URLSession+sharedMPTCP.swift:1]: Ah sorry, totally missed this. Let me double check to see if this is valid.

@Aperence did you try it locally or simply added it where it needs to get added?

Thanks again

Comment (Aperence) [reply firefox-ios/Client/Extensions/URLSession+sharedMPTCP.swift:1]: I built locally the application to make sure that it doesn’t generate new warnings.

Thanks again for the follow-up !

**Comment by Aperence** [comment]
Rebased it normally !

**Comment by Aperence** [comment]
Any news about this ?

File: firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift
Comment (OrlaM) [inline firefox-ios/Client/Extensions/URLSessionConfiguration+defaultMPTCP.swift:10]: This version of Firefox is only built for iOS and there are no plans to built for any other OS in the ecosystem so you can remove this macro.

**Review by OrlaM — CHANGES_REQUESTED** [review]
I'm surprised this isn't just default URLSession behaviour and it's very strange that Apple require an entitlement for it 🤔 
Great change either way, one minor nit otherwise LGTM

**Comment by Aperence** [comment]
Just applied the requested change. It probably isn't enabled by default because some time ago, some other options weren't (well) supported with MPTCP, and they probably never reconsidered up to now to enable it by default.

**Comment by nbhasin2** [comment]
Running BR and if all looks well I can merge it in 🤞🏼

**Review by nbhasin2 — APPROVED** [review]

**Review by OrlaM — APPROVED** [review]

**Comment by nbhasin2** [comment]
@Aperence might wanna rebase this one last time as you have approvals but there are certain number of warnings that are more on this PR compared to main. 

Thanks and once green I can hit merge

**Comment by Aperence** [comment]
Just rebased it now

**Pull request merged**
